feat(developer guide): New documentation for CI/CD (#125)
kestra-io/kestra.io: docs/developer-guide/cicd/github-action/README.md

---
order: 1
---

# GitHub

GitHub integrate a solution call [GitHub Actions](https://github.com/features/actions) that allow you
to create pipelines easily. Those pipelines are called `Workflows` and are build from `Actions` who
performs complex task with small amount of codes.

## Kestra Actions

> Kestra offer 2 `Actions` to create a CI/CD inside GitHub.

Reviewer: ```suggestion
Kestra offer 2 `Actions` to create a CI/CD pipeline within a GitHub repository.
```

> [Kestra Validate Action]() - Validate your flows and your templates before deploying anything.

Reviewer: ```suggestion
[Kestra Validate Action]() - Validates your flows and your templates.
```

Also, why did you add an empty link ?

> [Kestra Deploy Action]() - Deploy your flows and your templates to your Kestra server.

Reviewer: ```suggestion
[Kestra Deploy Action]() - Deploys your flows and your templates to your Kestra server.
```

## Examples

> Here is an example of a `Workflow` using `Kestra Actions` that validate all `Flows` and `Templates` before deploying them.

Reviewer: ```suggestion
Here is an example of a GitHub workflow using the Kestra actions to validate all `Flows` and `Templates` before deploying them.
```

Reviewer 2: +1

```yaml
name: Kestra CI/CD

on: [push]

jobs:
  # Start by validating all kestra resources
  Kestra_validate:
    runs-on: ubuntu-latest
    name: Kestra validate
    steps:

>       # Validate all templates within the client

Reviewer: ```suggestion
      # Validate all templates with the client
```

Reviewer 2: +1

      - name: Validate all templates locally
        uses: ./
        with:
          directory: ./kestra/templates
          resource: template
      # Validate all flows with the server
      - name: Validate all flows on server-side
        uses: ./
        with:
          directory: ./kestra/flows
          server: https://kestra.io/
  # If validation passed, deploy resources
  Kestra_deploy:
    runs-on: ubuntu-latest
    name: Kestra Deploy
    steps:
      # Deploy all templates in the io.kestra namespace
      - name: Deploy templates
        uses: ./
        with:
          namespace: io.kestra
          directory: ./kestra/templates
          resource: template
          server: https://kestra.io/

>       # Can only deploy in one namespace at once

Reviewer: ```suggestion
      # We can only deploy in one namespace at once,
```

Reviewer 2: +1

>       # so we have 2 differents step for our two namespaces io.kestra and io.kestra.dev

Reviewer: ```suggestion
      # so we have 2 different steps for our two namespaces io.kestra and io.kestra.dev
```

Reviewer 2: +1

One more change:
2 -> two

>       - name: Deploy prod flows

Reviewer: ```suggestion
      - name: Deploy flows in io.kestra
```

Reviewer 2: +1

>         description: Deploy flow in the io.kestra namespace

Reviewer: ```suggestion
        description: Deploy flows in the io.kestra namespace
```

Reviewer 2: +1

        uses: ./
        with:
          namespace: io.kestra
          directory: ./kestra/flows/prod
          server: https://kestra.io/
      - name: Deploy flows in io.kestra.dev

>         description: Deploy flow in the io.kestra.dev namespace

Reviewer: ```suggestion
        description: Deploy flows in the io.kestra.dev namespace
```

Reviewer 2: +1

---
order: 1
---

# GitHub

> GitHub integrates a solution called [GitHub Actions](https://github.com/features/actions) that allow you

Reviewer 2: ```suggestion
GitHub integrates a solution called [GitHub Actions](https://github.com/features/actions), allowing you to create CI/CD pipelines easily.
```

> to create CI/CD pipelines easily.

Reviewer 2: ```suggestion

```

> Those pipelines are called `Workflows` and are build from `Actions` which

Reviewer 2: ```suggestion
These pipelines are called Workflows and are built with Actions, which perform complex tasks with a small amount of code.
```

> performs complex tasks with small amount of codes.

Reviewer 2: ```suggestion

```

## Kestra Actions

> Kestra offer 2 `Actions` to create a CI/CD pipeline within a GitHub repository.

Reviewer 2: ```suggestion
Kestra offers two Actions to create a CI/CD pipeline within a GitHub repository.
```

> [Kestra Validate Action](https://github.com/kestra-io/validate-action/releases) - Validate your flows and your templates before deploying anything.

Reviewer 2: ```suggestion
* [Kestra Validate Action](https://github.com/kestra-io/validate-action/releases) - Validate your flows and templates before deploying anything.
```

> 

Reviewer 2: ```suggestion
```

> [Kestra Deploy Action](https://github.com/kestra-io/deploy-action/releases) - Deploy your flows and your templates to your Kestra server.

Reviewer 2: ```suggestion
* [Kestra Deploy Action](https://github.com/kestra-io/deploy-action/releases) - Deploy your flows and templates to your Kestra server.
```

## Examples

Here is an example of a `Workflow` using `Kestra Actions` that validate all `Flows` and `Templates` before deploying them.

::: warning

> `Flows` should always be deployed before `Templates`, to avoid flows running before their templates are created.

Reviewer 2: ```suggestion
Flows should always be deployed before Templates to avoid flows running before their templates are created.
```

:::

```yaml
name: Kestra CI/CD

on: [push]

jobs:
  # Start by validating all kestra resources
  Kestra_validate:
    runs-on: ubuntu-latest
    name: Kestra validate
    steps:
      # Validate all templates within the client
      - name: Validate all templates locally
        uses: ./
        with:
          directory: ./kestra/templates
          resource: template

>       # Validate all flows with the server

Reviewer 2: Same as Loïc's comment above.
